### PR TITLE
[5.4] Added ability to define custom class to use for pivot models.

### PR DESCRIPTION
**Pull request by pstephan1187**
Moved PR over from 5.2 branch: https://github.com/laravel/framework/pull/13304

### CODE REVIEW

**Comment by taylorotwell** [comment]
Do we have any way to get tests for this?

**Comment by pstephan1187** [comment]
I can work on writing some tests for it. I did already confirm that all tests written at the time of the request passed.

**Comment by morloderex** [comment]
This would be a welcome addition for me 👍

**Comment by JeremyNikolic** [comment]
I was hoping to get this feature for 5.3. Thanks for your work man

**Comment by Milewski** [comment]
I do use a lot of custom pivots... this will make my life easier ~

**Comment by pulasthibandara** [comment]
I work with some legacy databases with needs casting on model level. This would be really useful +1

File: src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php [hunk 1]
```diff
@@ -415,14 +435,31 @@ protected function getAliasedPivotColumns()
...
+     *
+     * @return array
+     */
+    protected function getPivotModelColumns()
+    {
+        $columns = [];
+
+        if ($this->using) {
+            $customPivotClass = $this->using;
+            $columns = $customPivotClass::getPivotColumns();
```
Comment (taylorotwell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: Why you have this new way of specifying which columns to load? Why not just use the traditional belongsToMany->with()?

**Comment by taylorotwell** [comment]
Is there any interest in continuing this PR?

**Comment by pstephan1187** [comment]
I do want to continue it. I just don't have the time to work on the tests at the moment.

**Comment by sisve** [comment]
Does this functionality differ from using "normal" models? Will they be called "pivot models" and work differently from existing models?

File: src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php (hunk 1 above)
Comment (pstephan1187) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: The reasoning behind this is that using the traditional method, you would need to defing the columns in every model using the custom pivot. So if you are doing a 3-way pivot, you would have to define the columns in up to 3 different models. That's bad duplication in my mind. The new method allows you to define the columns on the pivot model so you only have to define the columns in one sensible location.

**Comment by pstephan1187** [comment]
I've added a test. Tests are still new to me, so any direction on additional tests are welcome.

**Comment by pstephan1187** [comment]
@sisve Yes, they differ slightly in that you need to define which columns to load on the pivot model: [13304](https://github.com/laravel/framework/pull/13304)

**Comment by GrahamCampbell** [comment]
Please rebase this on our master. :)

**Comment by pstephan1187** [comment]
@GrahamCampbell Changes have been rebased.

**Comment by taylorotwell** [comment]
I merged this but removed some things to simplify it for now so we can just get the basic spirit of it in. I wanted to get it in the framework at least as a simple starting point.

I removed the automatic eager loading from the constructor because I don't generally think database queries should be done just to new up an object. It makes it hard to new up something for testing, etc.

I also removed the ability to define the attributes to load on the model itself mainly because I feel that is fairly easily recreated by just putting a public static property on the custom pivot model and doing: `->withPivot(PivotModel::$with)` or something to that effect. You still aren't having to repeat the actual column names.

Thanks!

**Comment by nelson6e65** [comment]
By the way... 

> "[...] is fairly easily recreated by just putting a public static property on the custom pivot model and doing: `->withPivot(PivotModel::$with)` or [...] " @taylorotwell 

...you **can't** use the `static $with` property; There is already a non static `$with` propery defined in `Illuminate\Database\Eloquent\Relations\Pivot`, for other porpoise.

> "Cannot redeclare non static Illuminate\Database\Eloquent\Relations\Pivot::$with as static [...]"

**Pull request merged**